From 38c86ca862a8cccb8ab0590d8a4f309991d7a5e9 Mon Sep 17 00:00:00 2001
From: Geoff Thorpe <Geoff.Thorpe@freescale.com>
Date: Wed, 11 Nov 2009 21:26:57 -0500
Subject: [PATCH] qman: fix hot-potato test, and reduce FQ requirement.

We were not releasing allocated FQIDs, so fix that. Also, to avoid
exhaustion of the FQ pool, use fewer hot-potato handlers per cpu.

Signed-off-by: Geoff Thorpe <Geoff.Thorpe@freescale.com>
[Kevin: Original patch taken from Freescale p4080 SDK 2.0.1 ISO
image. Apply cleanly to kernel 2.6.27]
Integrated-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/hwqueue/qman_test_hotpotato.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/drivers/hwqueue/qman_test_hotpotato.c b/drivers/hwqueue/qman_test_hotpotato.c
index 017e8b0..f727570 100644
--- a/drivers/hwqueue/qman_test_hotpotato.c
+++ b/drivers/hwqueue/qman_test_hotpotato.c
@@ -139,7 +139,7 @@ static dma_addr_t frame_dma;
 /* the main function waits on this */
 static DECLARE_WAIT_QUEUE_HEAD(queue);
 
-#define HP_PER_CPU 	16
+#define HP_PER_CPU 	2
 #define HP_LOOPS	8
 /* 80 bytes, like a small ethernet frame, and bleeds into a second cacheline */
 #define HP_NUM_WORDS	80
@@ -269,6 +269,7 @@ static void destroy_per_cpu_handlers(void *ignore)
 			panic("qman_oos_fq(rx) failed");
 		qman_destroy_fq(&handler->rx, 0);
 		qman_destroy_fq(&handler->tx, 0);
+		qm_fq_free(handler->fqid_rx);
 		list_del(&handler->node);
 		kmem_cache_free(hp_handler_slab, handler);
 	}
-- 
1.6.0.4

